src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+    }
+
+    CreateIndexRequest request = new CreateIndexRequest(index);
+    callWithRetries(
+        "create index " + index,
+        () -> {
+          try {
+            client.indices().create(request, RequestOptions.DEFAULT);
+          } catch (ElasticsearchStatusException | IOException e) {
+            if (!e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)) {
+              throw e;
+            }
+          }
+
+          return null;
+        }
+    );
+  }
+
+  /**
+   * Creates a mapping for the given index and schema.
+   *
+   * @param index the index to create the mapping for
+   * @param schema the schema to map
+   */
+  public void createMapping(String index, Schema schema) {
+    PutMappingRequest request = new PutMappingRequest(index).source(Mapping.buildMapping(schema));
+    callWithRetries(
+        "create mapping",
+        () -> client.indices().putMapping(request, RequestOptions.DEFAULT)
+    );
+  }
+
+  /**
+   * Flushes any buffered records.
+   */
+  public void flush() {
+    bulkProcessor.flush();
+  }
+
+  /**
+   * Checks whether the index already has a mapping or not.
+   * @param index the index to check
+   * @return true if a mapping exists, false if it does not
+   */
+  public boolean hasMapping(String index) {
+    MappingMetaData mapping = getMapping(index);
+    return mapping != null && !mapping.sourceAsMap().isEmpty();
+  }
+
+  /**
+   * Buffers a record to index. Will ensure that there are no concurrent requests for the same
+   * document id.
+   *
+   * @param record the record to index
+   * @param request the associated request to send
+   * @throws ConnectException if one of the requests failed
+   */
+  public void index(SinkRecord record, DocWriteRequest<?> request) {
+    if (isFailed()) {
+      close();
+      throw error.get();
+    }
+
+    if (recordMap != null) {
+      // make sure that only unique records are being sent in a batch
+      while (recordMap.contains(request.id())) {
+        flush();
+        try {
+          Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+        } catch (InterruptedException e)  {
+          throw new ConnectException(e);
+        }
+      }
+    }
+
+    addToRecordMap(request.id(), record);
+    bulkProcessor.add(request);
+  }
 
-public interface ElasticsearchClient extends AutoCloseable {
+  /**
+   * Checks whether the index exists.
+   *
+   * @param index the index to check
+   * @return true if it exists, false if it does not
+   */
+  public boolean indexExists(String index) {
+    GetIndexRequest request = new GetIndexRequest(index);
+    return callWithRetries(
+        "check if index " + index + " exists",
+        () -> client.indices().exists(request, RequestOptions.DEFAULT)
+    );
+  }
 
-  enum Version {
-    ES_V1, ES_V2, ES_V5, ES_V6, ES_V7
+  /**
+   * Maps a record to the document id.
+   * @param id the document id
+   * @param record the record
+   */
+  private void addToRecordMap(String id, SinkRecord record) {
```

Comment:
leftover from `BulkProcessor` implementation, while yes it is only called once, given it is wrapped in a `null` check it is probably better to abstract it away in case of future use